# Objective:

Can understanding that programming is basically manipulating data to fit into a specific requirement.

# What is programming?

![image.png](https://storage.googleapis.com/rg-ai-bootcamp/python-prep/programming-main-focus-min.png)

The real foundation of programming, the essence of it is to manipulate data to adhere to what we need to do next.

This "what we need to do next" itself is depends on what kind of programming we currently doing. For example, if we're doing a web programming, the "thing we need to do next after manipulating data" might be to render a web page, saving user data to database, or sending an email. If we're doing a game programming, what we need to do might be to render a game screen, playing a sound, or saving the game state.

Let's check below simple program:

In [1]:
#@title Preparation so we can do get_emotion(data), you can just run and ignore this line of code for now
%pip install transformers
%pip install torch
from transformers import pipeline

classifier = pipeline(task="text-classification", model="SamLowe/roberta-base-go_emotions", top_k=None)

def get_emotion(text):
    model_outputs = classifier(text)
    return highest_confidence_label(model_outputs[0])

def highest_confidence_label(data):
    highest_score = max(data, key=lambda x:x['score'])
    confidence = round(highest_score['score'] * 100, 2)
    return f"{highest_score['label']}, {confidence}% confidence"

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/1.92k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/499M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/380 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/798k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.11M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/280 [00:00<?, ?B/s]

In [2]:
get_emotion("I'm so excited to meet you guys there!")

'excitement, 81.15% confidence'

In above line of code we can see that:

- We have a data that we want to process: "I'm so excited to meet you guys there!"
- We have a requirement that we want to do: Get the emotion probability of the data using get_emotion("we put our data here")
- And it return a response of the confidence of the emotion probability: excitement, 81.15% confidence

So we have an input, we process it, and we get an output, we can use that output as an input again, and so on. That's the essence of programming.

> Please have fun: While you're at it, have fun change the data above (change inside the `"` symbol) to feed it with your own data, and see what kind of emotion probability that you get. Try to change it something like "Why are you running late?", "Stuck in traffic, ugh!", "I'm not feeling well today, sorry guys"

# Challenge

## Challenge 1

For the first challenge we'll start with a really simple task:

Change the `get_emotion("I'm so excited to meet you guys there!")` to "I'm really stressed out for this meeting, I hope it goes well" and send the result of the probability.

In [5]:
get_emotion("I'm so happy to meet you guys there!") #Change the text inside get_emotion()

'joy, 88.13% confidence'

To send the result:
1. Input your student_id and name in the box below
2. Change the text "my result" below with the text output from get_emotion() above. example: "excitement, 81.15% confidence"
3. Run the code block by pressing the play button.

In [18]:
!pip install rggrader

from rggrader import submit

# @title #### Student Identity
student_id = "REA6UCWBO" # @param {type:"string"}
name = "Ida Bagus Teguh Teja Murti" # @param {type:"string"}

result = "done"

# Submit Method
assignment_id = "001_the-basics"
question_id = "01_get_emotion"
submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

# Uh oh, the input changed!

As we said above, one of the most fundamental thing for us to understand programming is to manipulate data. For above code, we have a convenient text that we can just pass to `get_emotion`, but what if the data isn't that convenient? Such as we have two inputs like so:

```
"I'm so {emotion} to meet you guys there!"
"excited"
```

We have two separate inputs, and, to make sure that we can process that input to `get_emotion`, we need to make sure that input is compatible with `get_emotion` requirement. Which can be done like so:

In [7]:
get_emotion("I'm so {emotion} to meet you guys there!".replace("{emotion}", "happy"))

'joy, 88.13% confidence'

So now we're talking! If we see above, we'll see additional code after our "I'm so {emotion} to meet you guys there!" data, which is `.replace("{emotion}", "excited")`.

We'll learn about how that works later, but for now, let's just say that we're replacing the `{emotion}` part of the data with the "excited" data, as the `get_emotion` function requires to have only a single text data as an input.

# Conclusion

As we said above, the complexity on programming mostly rely only on making sure that your data is ready to be inputted to our given task. When we want to save our data to database, we want to manipulate our data so it can be saved properly. If we want to render a web page, we want to manipulate our data so it can be rendered properly. And so on.

It's simple right?

# Terminology alert!

Let's return to our first line of code:
    
```
get_emotion("I'm so excited to meet you guys there!")
```

- `get_emotion` is a function
- `"I'm so excited to meet you guys there!"` is a data with a data type of string

We'll learn more about function and data type later, and other terminology that's related to function and data type, but for now just remember above terminology.

# Data Types and comments

We've already learned this before:

- `"I'm so excited to meet you guys there!"` is a data with a data type of string

There are several data types in Python, we'll explore three of them for now: `string`, `float` and `int`.

# Google Colab, Jupyter Notebook, and Visual Studio Code will "print" the last line of code automatically

Before we continue, please check below code

In [8]:
"Hi everyone"
"Hello!"

'Hello!'

As you can see above, our result to above two lines of code will only show the result of the last line of code. This concept is called printing, we can print any code in any line by using `print()` like below:

In [9]:
"Yes!"
print("Hello!")
"Hi!"

Hello!


'Hi!'

As you can see above we have 3 lines of code, but we only print the second and third line of code. The second line of code is printed because we use `print()`, while the third line of code is printed because it is the last line of code.

# Challenge!

## Challenge 1

Make sure both of the string below are printed

In [10]:
print("Halo dunia")
"Hi world"

Halo dunia


'Hi world'

When you are done with the above challenge, then:
1. Input your student_id and name in the box below
2. Change the text "my result" below with the text "done"
3. Run the code block by pressing the play button.

In [17]:
!pip install rggrader

from rggrader import submit

# @title #### Student Identity
student_id = "REA6UCWBO" # @param {type:"string"}
name = "Ida Bagus Teguh Teja Murti" # @param {type:"string"}

result = "done"

# Submit Method
assignment_id = "002_data_types"
question_id = "01_using_print"
submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

# String

A string is a data type that is used to represent text. It is enclosed by either single or double quotes.

```python
'Thank you for coming to my TED talk'

"I'm so excited to meet you guys there!"
```

A string can be manipulated like what we've already seen in previous learning

```python
"I'm so {emotion} to meet you guys there!".replace("{emotion}", "excited")
```

We see above our more complex string from previous learning: `"I'm so excited to meet you guys there!"`, then we replace the `{emotion}` with `"excited"`.

# Arguments

We'll learn more about method and a concept that's basically similar to method called function later, but for now, let's give a quick brief about methods and arguments.

If we see the code below:

```python
"I'm so {emotion} to meet you guys there!".replace("{emotion}", "excited")
```

What's called method is the `.replace()` part, while the `{emotion}` and `"excited"` are called arguments.

The number of arguments itself can be varied between method, sometimes a method can have no argument, sometimes it can have one, two, or even more arguments.

# String methods

## `.lower()`

This method will convert all characters in a string into lowercase

In [12]:
"I'M NOT SHOUTING".lower()

"i'm not shouting"

## `.strip()`

This method will remove all leading and trailing whitespaces from a string

In [13]:
"   Hello   ".strip()

'Hello'

## `.replace()`

This method will replace all occurrences of a string with another string

In [14]:
"Bob is a cat".replace("cat", "dog")

'Bob is a dog'

If we add an empty string as the second argument, it will remove all occurrences of a string

In [15]:
"cat cat cat dog human horse".replace("cat", "")

'   dog human horse'

We won't list down every methods here, you can check more on https://www.w3schools.com/python/python_ref_string.asp

# Challenge!

## Upper

Using `.upper()` method, convert the string below into uppercase

In [16]:
"hi my name is andi".upper()

'HI MY NAME IS ANDI'

When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [19]:
# Submit Method
assignment_id = "002_data_types"
question_id = "02_upper"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

## Replace

Replace below name from "Ridho" to "Fajar"!

In [21]:
"Hi, my name is Ridho".replace("Ridho", "Fajar")

'Hi, my name is Fajar'

When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [22]:
# Submit Method
assignment_id = "002_data_types"
question_id = "03_replace"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

## String

Remove all leading and trailing whitespaces from below string

In [23]:
"   HELLO   ".strip()

'HELLO'

When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [24]:
# Submit Method
assignment_id = "002_data_types"
question_id = "04_string"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

## Method chaining

We can chain multiple methods together, for example:

```python
"I'm so excited to meet you guys there!".lower().replace("excited", "happy")
```

Will result in:

```python
"i'm so happy to meet you guys there!"
```

# Challenge

## Challenge 1

Replace all occurrences of "cat" to "" (empty string) and remove all leading whitespaces (the output should be: "dog human horse")

In [26]:
"cat cat cat dog human horse".replace('cat','').strip()

'dog human horse'

When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [27]:
# Submit Method
assignment_id = "002_data_types"
question_id = "05_method_chain_1"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

## Challenge 2

Convert below string to "IT'S SO COOL!"

In [30]:
print("  It's so cool!    ".upper().strip())

IT'S SO COOL!


When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [31]:
# Submit Method
assignment_id = "002_data_types"
question_id = "06_method_chain_2"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

## Challenge 3

Convert below string to "I'm not a bird"

Note:

- If you want to delete a word from a sentence, make sure you add space before and after the word so it will remove the word and the space before and after it.

```python
"I'm so hungry".replace("so", "") # WRONG, it will result in "I'm  hungry" (notice the extra space)
"I'm so hungry".replace(" so ", "") # CORRECT, it will result in "I'm hungry"
```

In [32]:
"I'm not a cat"

"I'm not a cat"

When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [33]:
# Submit Method
assignment_id = "002_data_types"
question_id = "07_method_chain_3"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

# Comment

Comment is a text that is ignored by the computer, it's mostly used to explain what a code does.

In [34]:
# This is a comment, this line will not be executed
# def my_function():
#     print("Hello from a function")
# above code will not be executed either

"This line will be executed" # print("This line will not be executed")

'This line will be executed'

# String concatenation

We can concatenate (combine) two strings together using `+` operator

In [35]:
"Hello " + "world!"

'Hello world!'

# Challenge

## Challenge 1

Try to concatenate below string with Rudi's name

In [36]:
"Imam, Andi, Ridho" # Convert to "Imam, Andi, Ridho, Rudi"

'Imam, Andi, Ridho'

When you are done with the above challenge, then:

1. Change the text "my result" below with the text "done"
2. Run the code block by pressing the play button.

In [38]:
# Submit Method
assignment_id = "002_data_types"
question_id = "08_concat"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

# Numbers (integers and floats)

Numbers in Python are divided into two types: integers and floats. Integers are whole numbers, while floats are decimal numbers.

Below are integers:

```python
1
30
70
```

Below are floats:

```python
1.0
30.0
70.0
```

We can do basic math operations with either integers or floats:

In [39]:
(5 + 3) * 2.0

16.0

# Challenge!

## Challenge 1

Create a formula of 10 times 2.
The output = 20

In [40]:
# Write the code
10 * 2

20

When you are done with the above challenge, then:

1. Change the text "my result" below with the formula you created above.
2. Run the code block by pressing the play button.

In [41]:
# Submit Method
assignment_id = "002_data_types"
question_id = "09_math_1"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'

## Challenge 2

Create a formula of 10 plus 2 then times 3 (remember the order of operations, you might need to use parentheses?)
The output = 36

In [42]:
# Write the code
(10 + 2 ) * 3

36

When you are done with the above challenge, then:

1. Change the text "my result" below with the formula you created above.
2. Run the code block by pressing the play button.

In [43]:
# Submit Method
assignment_id = "002_data_types"
question_id = "10_math_2"

result = "done"

submit(student_id, name, assignment_id, result, question_id)

'Assignment successfully submitted'